File: Leetcode/Minimum Insertion Steps to Make a String Palindrome/Minimum Insertion Steps to Make a String Palindrome.java
class Solution {
    public int minInsertions(String s) {
        //Finding the size of given string
        int n = s.length();
        //Initialising dp array. It will represent the longest common subsequence between first i characters of first string and first j characters of second string
        int[][] dp = new int[n+1][n+1];
        
        //Looping through start and end of the string. Thus dp will consider the string from start and string from end and then store the ans
        for (int i = 0; i < n; ++i){
          for (int j = 0; j < n; ++j){
              
              //If both the characters are equal, then we increment the previous dp value otherwise we take max of the next character considered for both strings
              dp[i + 1][j + 1] = s.charAt(i) == s.charAt(n - 1 - j) ? dp[i][j] + 1 : Math.max(dp[i][j + 1], dp[i + 1][j]);
          }  
        }
        //Returning ans
        return n - dp[n][n];
    }
}

File: Leetcode/Minimum Insertion Steps to Make a String Palindrome/Minimum Insertion Steps to Make a String Palindrome.py
class Solution:
    def minInsertions(self, s: str) -> int:
    
        m=len(s)          # Calculating the length of the string
        
        t=[[0 for i in range(m+1)] for i in range(m+1)]
        r=s[::-1]         # Reversing the string
        for i in range(1,m+1):
            for j in range(1,m+1):           # Evaluating result by iterating through every character twice
                if m==0:
                    t[i][j]=0
                elif s[i-1]==r[j-1]:
                    t[i][j]=t[i-1][j-1]+1
                else:
                    t[i][j]=max(t[i-1][j],t[i][j-1])
        return m-t[m][m]  # Returning result
